deepdrivemd/agents/stream/dbscan.py

```diff
@@ -0,0 +1,363 @@
+    tmp_dir = f"{top_dir}/tmp"
+    published_dir = f"{top_dir}/published_outliers"
+
+    if(not os.path.exists(tmp_dir)):
+         os.mkdir(tmp_dir)
+    if(not os.path.exists(published_dir)):
+         os.mkdir(published_dir)
+    return top_dir, tmp_dir, published_dir
+
+def predict(cfg, model_path, cvae_input, batch_size=32):
+    cvae = build_model(cfg, model_path)
+    input = np.expand_dims(cvae_input[0], axis = -1)
+
+    cm_predict = cvae.return_embeddings(input, batch_size)
+    del cvae 
+    K.clear_session()
+    return cm_predict
+
+def outliers_from_latent(cm_predict, eps=0.35, min_samples=10):
+    cm_predict = cp.asarray(cm_predict)
+    db = DBSCAN(eps=eps, min_samples=min_samples, max_mbytes_per_batch=100).fit(cm_predict)
+    db_label = db.labels_.to_array()
+    print("unique labels = ", np.unique(db_label))
+    outlier_list = np.where(db_label == -1)
+    K.clear_session()
+    return outlier_list
+
+def cluster(cfg, cm_predict, outlier_list, eps, min_samples):
+    outlier_count = cfg.outlier_count
+    while outlier_count > 0:
+        n_outlier = 0
+        try:
+            outliers = np.squeeze(outliers_from_latent(cm_predict, eps=eps, min_samples=min_samples)) 
+            n_outlier = len(outliers)
+        except Exception as e:
+            print(e)
+            print("No outliers found")
+
+        print(f'eps = {eps}, min_samples = {min_samples}, number of outlier found: {n_outlier}')
+
+        if n_outlier > cfg.outlier_max: 
+            eps = eps + 0.09*random.random()
+            min_samples -= int(random.random() < 0.5)
+            min_samples = max(5, min_samples)
+        elif n_outlier < cfg.outlier_min:
+            eps = max(0.01, eps - 0.09*random.random())
+            min_samples += int(random.random() < 0.5)
+        else: 
+            outlier_list.append(outliers) 
+            break 
+        outlier_count -= 1
+    return eps, min_samples
+
+
+def write_pdb_frame(frame, original_pdb, output_pdb_fn):
+    pdb = PDBFile(str(original_pdb))
+    f = open(str(output_pdb_fn), 'w')
+    PDBFile.writeFile(pdb.getTopology(), frame, f)
+    f.close()
+
+def write_top_outliers(cfg, tmp_dir, top):
+    positions = top[0]
+    velocities = top[1]
+    md5s = top[2]
+
+    for p,v,m in zip(positions, velocities, md5s):
+        outlier_pdb_file = f'{tmp_dir}/{m}.pdb'
+        outlier_v_file = f'{tmp_dir}/{m}.npy'
+        write_pdb_frame(p, cfg.init_pdb_file, outlier_pdb_file)
+        np.save(outlier_v_file, v)
+
+def write_db(top, tmp_dir):
+    outlier_db_fn = f'{tmp_dir}/OutlierDB.pickle'
+    outlier_files = list(map(lambda x: f'{tmp_dir}/{x}.pdb', top[2]))
+    rmsds = top[3]
+    db = OutlierDB(tmp_dir, list(zip(rmsds, outlier_files)))
+    with open(outlier_db_fn, 'wb') as f:
+        pickle.dump(db, f)    
+    return db
+
+def publish(tmp_dir, published_dir):
+    dbfn = f"{published_dir}/OutlierDB.pickle"
+    subprocess.getstatusoutput(f"touch {dbfn}")
+
+    mylock = LockFile(dbfn)
```

Comment:
Instead of using FileLock, they recommend to use another package such as fasteners since this one is depreciated. I also read a PR saying there are potential issues with cleaning lock files on failure.

Depreciated: https://pypi.org/project/lockfile/

Potential issue: https://github.com/openai/mujoco-py/issues/424

Fasteners link: https://fasteners.readthedocs.io/en/latest/examples.html

Reply:
I have been using FileLock for a long time and so far it worked fine, I never had any problems. If the program crashes, locks might indeed remain but I would clean everything before restarting the program anyway.

Reply:
Ok that is fine for now then. In the future we will want to update it though so the library is supported.